cdap-integration-test/src/main/java/co/cask/cdap/test/IntegrationTestBase.java

```diff
@@ -53,29 +59,110 @@
 
 import java.io.File;
 import java.io.IOException;
 import java.net.URI;
+import java.net.URL;
 import java.util.List;
 import java.util.Map;
+import java.util.Properties;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
 import javax.annotation.Nullable;
+import javax.ws.rs.ServiceUnavailableException;
 
 /**
  *
  */
 public class IntegrationTestBase {
+  private static final Logger LOG = LoggerFactory.getLogger(IntegrationTestBase.class);
 
   @ClassRule
   public static final TemporaryFolder TEMP_FOLDER = new TemporaryFolder();
 
   @Before
   public void setUp() throws Exception {
+    checkSystemServices();
+    assertUnrecoverableResetEnabled();
     assertIsClear();
   }
 
+  private void checkSystemServices() throws ServiceUnavailableException {
+    MonitorClient monitorClient = getMonitorClient();
+
+    try {
+      checkSystemServicesOk(monitorClient);
+    } catch (UnauthorizedException e) {
+      LOG.info("Caught UnAuthorizedException, Getting Credentials");
+      // security is enabled, we need to get access token before checking system services
+      URL url = null;
+      try {
+        Properties properties = new Properties();
+        properties.setProperty("security.auth.client.username", System.getProperty("userName"));
+        properties.setProperty("security.auth.client.password", System.getProperty("password"));
+        AuthenticationClient authClient = new BasicAuthenticationClient();
+        authClient.configure(properties);
+        ConnectionConfig connectionConfig = getClientConfig().getConnectionConfig();
+        authClient.setConnectionInfo(connectionConfig.getHostname(), connectionConfig.getPort(), false);
+        AccessToken accessToken = getAccessToken(authClient);
+        System.setProperty("accessToken", accessToken.getValue());
+        System.setProperty("accessTokenType", accessToken.getTokenType());
+        System.setProperty("expiresIn", String.valueOf(accessToken.getExpiresIn()));
+        checkSystemServicesOk(getMonitorClient());
+      } catch (Exception ex) {
+        throw Throwables.propagate(ex);
+      }
+    }
+    LOG.info("CDAP Services are up and running!");
+  }
+
+  private AccessToken getAccessToken(AuthenticationClient authClient) {
+    long authClientRetry = TimeUnit.MINUTES.toSeconds(3);
+    int numTries = 0;
+    do {
+      try {
+        numTries++;
+        TimeUnit.SECONDS.sleep(1);
+        return authClient.getAccessToken();
+      } catch (InterruptedException e) {
+        // no-op
+      } catch (IOException e) {
+        // no-op
+      }
+    } while (numTries <= authClientRetry);
+    throw new ServiceUnavailableException("Unable to connect to AuthServer to obtain AccessToken");
+  }
+
+  private void checkSystemServicesOk(MonitorClient monitorClient) throws UnauthorizedException {
+    // we will wait for upto 10minutes for CDAP services to be up.
+    long maxRetries = TimeUnit.MINUTES.toSeconds(10);
```

Comment:
We should consolidate all retry time into one and make it a constant. Maybe call it `MAX_REST_WAIT_TIME`, or something.